torch/csrc/perf_tools/ThroughputBenchmark-inl.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include <random>
+#include <thread>
+
+#include <torch/csrc/utils/pybind.h>
+#include <torch/csrc/jit/pybind_utils.h>
+
+namespace torch {
+namespace throughput_benchmark {
+namespace detail {
+
+template <class Input, class Output, class Model>
+BenchmarkExecutionStats BenchmarkHelper<Input, Output, Model>::benchmark(
+    const BenchmarkConfig& config) const {
+  CHECK(initialized_);
+  TORCH_CHECK(
+      config.num_worker_threads == 1,
+      "Only parallelization by callers is supported");
+
+  // We pre-generate inputs for each of the threads and avoid threadsafety
+  // issues for random calls. As thread safe random is not provided in the
+  // standard library. We also replicate IValues. So we don't need to copy
+  // them during the main benchmark phase.
+  std::vector<std::vector<Input>> thread_inputs(config.num_calling_threads);
+  std::vector<size_t> input_iters(config.num_calling_threads);
+  {
+    std::random_device seeder;
+    std::mt19937 engine(seeder());
+    TORCH_CHECK(!inputs_.empty(), "Please provide benchmark inptus");
```

Comment:
"Please provide benchmark inputs" is not a helpful error message, i think.
Also say what was provided. Maybe: "inputs was expected to be non-zero number of elements, but got inputs to be zero elements."

Reply:
This message will fire in a case user didn't call num_input at all. This is the only way to provide inputs, you can not explicitly pass an empty array with the current API. 
I can change this message to talk about add_input() call not being made.